gammapy/catalog/core.py

```diff
@@ -257,18 +258,44 @@ def _make_source_object(self, index):
             name_extended = data["Source_Name"].strip()
         else:
             name_extended = None
+
+        if self._source_name_key in data and self._source_name_key == "PSR_Name":
+            name_spectral = data[self._source_name_key].strip()
+        elif self._source_name_key in data and self._source_name_key == "PSRJ":
+            name_spectral = f"PSR{data[self._source_name_key].strip()}"
+        elif "Source_name" in data:
+            name_spectral = data["Source_name"].strip()
+        else:
+            name_spectral = None
+
         try:
             idx = self._lookup_extended_source_idx[name_extended]
             data_extended = table_row_to_dict(self.extended_sources_table[idx])
         except (KeyError, AttributeError):
             data_extended = None
 
-        source = self.source_object_class(data, data_extended)
+        try:
+            idx = self._lookup_spectral_source_idx[name_spectral]
+            data_spectral = table_row_to_dict(self.spectral_table[idx])
+        except (KeyError, AttributeError):
+            data_spectral = None
+
+        source = self.source_object_class(data, data_extended, data_spectral)
         return source
 
+    @lazyproperty
+    def _lookup_spectral_source_idx(self):
+        if self._source_name_key == "PSRJ":
+            source_name_key = "NickName"
+        else:
+            source_name_key = self._source_name_key
```

Comment:
maybe move this to a separate private function
`_get_source_name_key()` that return `self._source_name_key` here
 and re-write it in the subclass to return `NickName`
 (and why _source_name_key is different  from  source_name_key in that case ?)

Reply:
The difference is because in the 3PC, the column that matches the column "PSRJ" of the main table in the spectral table is `Nickname`, while it is the same in the 2PC.